### PR TITLE
Reduce toot size limit for language detection

### PR DESCRIPTION
**Pull request by stevenroose**
Improves https://github.com/tootsuite/mastodon/issues/9550

### CODE REVIEW

**Comment by ThibG** [comment]
The issue is that automatic detection for such short texts is completely unreliable :/

**Comment by faho** [comment]
>The issue is that automatic detection for such short texts is completely unreliable :/

Is there any data about that?

Because, anecdotally, 140 is long enough to basically be useless, and google translate at least (no idea if that actually uses cld3 in the background) can detect the language for much shorter texts.

**Comment by Gargron** [comment]
Google Translate doesn't use CLD3, they use machine learning and it's trained on a very large dataset. Not something you can replicate on a small standalone mastodon server.

**Comment by faho** [comment]
>Google Translate doesn't use CLD3

Sure, but that shows that it's _possible_ to reliably detect languages with texts shorter than 140 characters.

So I'd like to know if anyone knows why 140 was picked, and if it's maybe possible that CLD3 could handle shorter texts?

Maybe 20 is too short, but 40 or 60 would work?

**Comment by Gargron** [comment]
140 was picked arbitrarily after some back and forth testing, as documented in the screenshot here: #8010. I'm not claiming it's a scientific number, but the less characters there are, the more wildly inaccurate CLD3 is.

**Comment by faho** [comment]
>140 was picked arbitrarily after some back and forth testing, as documented in the screenshot here: 

That screenshot shows:

- It's worthless for single words - which is obvious, since e.g. "die" is both english and german ("die bart, die"), and that'll happen frequently.

- It gives you a "probability" score (which I'm guessing is also used for the "reliable?" bool?) that is worthless for single words, but not all that bad for longer sentences? For the two longer tests, it's < 50% (and the longest is a correct guess). These are 51/24 characters, which is a far cry from the 140 that is used currently!

So it doesn't seem all that awful to use a sliding scale? I.e. for things under 40 characters or things marked unreliable, don't trust the detection.

**Comment by nightpool** [comment]
If i remember correctly, CLD used to contain a disclaimer that it wasn't supposed to be used on texts smaller then N size (i believe around 300-500). We're using it with 140 but I wouldn't want to go any lowers.

**Comment by stevenroose** [comment]
Are there any alternative language detection systems?

Of course, I still consider #9550 to be the better way to go here.

**Review by Gargron — CHANGES_REQUESTED** [review]

File: app/lib/language_detector.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@
 class LanguageDetector
   include Singleton
 
-  CHARACTER_THRESHOLD = 140
+  CHARACTER_THRESHOLD = 20
```
[the remark this answers is not in the document]
Comment (stevenroose) [reply app/lib/language_detector.rb]: I don't see why a language detector wouldn't be able to correctly classify toots like `Yes!`, `Good idea.`, `Je ne sais pas.`, `Donde esta la bilbioteca?` etc with a high reliability.

[the remark this answers is not in the document]
Comment (Gargron) [reply app/lib/language_detector.rb]: Sorry for a late response. But you don't need to believe my word, just play around with the detector and see how reliable it is for those phrases.

grafik

**Pull request closed without merging**